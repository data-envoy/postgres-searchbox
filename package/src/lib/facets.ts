import format from 'pg-format';
import type { Props, GetFacetsReturn } from './facets.types.js';

/**
 * Ambiguous params
 * maxValuesPerFacet: The maximum number of facet values to return for each facet in a regular search.
 * sortFacetValuesBy: Controls how facet values are fetched.
 * maxFacetHits: Maximum number of facet hits to return during a search for facet values.
 */

export const getFacets = async ({
  facets: facetsMaybeString,
  attributesForFaceting,
  maxValuesPerFacet,
  sortFacetValuesBy,
  maxFacetHits,
  renderingContent,
}: Props): Promise<GetFacetsReturn> => {
  // Build facets based on 2 conditionals
  const facets: readonly string[] =
    facetsMaybeString?.includes('*') || !facetsMaybeString
      ? attributesForFaceting // facetsMaybeString is * or undefined
      : Array.isArray(facetsMaybeString) // is it an array already?
      ? facetsMaybeString // yes
      : [facetsMaybeString]; // no, string cast to array

  if (!facets?.length) return null;

  const limit = Number.isInteger(maxValuesPerFacet) ? maxValuesPerFacet : 10;

  // For things like
  const cte = facets
    .map((facet) =>
      format(
        /* sql */ `
        %I AS (
          SELECT coalesce( 
            json_object_agg( %s, cnt ORDER BY cnt DESC ), 
            '{}'::json
          ) as details
            FROM (
              SELECT %s, count(*) AS cnt
              FROM all_selection
              WHERE %s IS NOT NULL
              GROUP by %s
              ORDER BY ${
                sortFacetValuesBy === 'count' ? `count(%L) DESC` : `%s ASC`
              } 
              LIMIT ${limit}
            ) t
        )
      `,
        `${facet}_selection`,
        facet,
        facet,
        facet,
        facet,
        facet,
        facet
      )
    )
    .join(',');

  const json = /* sql */ `
    'facets', json_build_object(
      ${facets
        .map((facet) =>
          format(`%L, (SELECT details FROM %I)`, facet, `${facet}_selection`)
        )
        .join(',')}
    )`;

  const order = renderingContent?.facetOrdering?.facets?.order;
  const values = renderingContent?.facetOrdering?.values;

  // Build up the object based on the facets
  let newOrder: string[] | undefined;
  if (order?.includes('*')) {
    newOrder = facets.map((f) => f);
  }

  return {
    db: { cte, json },
    renderingContent: {
      ...renderingContent,
      facetOrdering: {
        ...renderingContent?.facetOrdering,
        facets: {
          order: newOrder || order,
        },
        // TODO values
        values,
      },
    },
  };
};
