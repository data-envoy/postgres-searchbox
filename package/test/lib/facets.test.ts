import { getFacets } from '@/lib/facets.js';

/**
 * Test attributes
 */

describe('attributes', () => {
  it.only('should return SQL from facetFilters', async () => {
    const returnValue = await getFacets(
      {
        facetFilters: `[
          ["attribute1:value", "attribute1:value2", "attribute2:value", "attribute5:value" ],
          "attribute3:value" , "attribute3:-value2", "attribute4:value2"
        ]`,
        facets: 'attribute1',
        numericFilters: `["price>=11","price>=12","price<=15","price=[12,14]"]`,
      },
      {
        facets: ['attribute3', 'price'],
        disjunctiveFacets: ['attribute1', 'attribute2'],
        hierarchicalFacets: [],
      }
    );

    expect(returnValue?.db?.formatted).toMatchSnapshot();
  });
});
