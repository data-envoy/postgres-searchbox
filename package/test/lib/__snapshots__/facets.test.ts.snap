// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facets should return workingSQL fragments 1`] = `
{
  "Abshire - Carroll": 2,
  "Hayes - Kutch": 2,
  "Hickle - Bins": 2,
  "Kessler Inc": 2,
  "Langworth, Kuvalis and Kohler": 2,
  "Littel, Borer and Bergstrom": 2,
  "Murphy and Sons": 2,
  "O'Keefe, Ullrich and Heller": 2,
  "Powlowski, Brekke and Hamill": 2,
  "Smitham Group": 2,
}
`;

exports[`facets should return workingSQL fragments 2`] = `
{
  "12": 7,
  "13": 6,
  "14": 5,
  "19": 4,
  "22": 4,
  "24": 6,
  "27": 5,
  "29": 4,
  "31": 4,
  "38": 4,
}
`;

exports[`facets should return workingSQL fragments 3`] = `
{
  "avg": 23.81,
  "max": 40,
  "min": 10,
  "sum": 2381,
}
`;
